drivers/can/mcp2515.c

```diff
@@ -0,0 +1,748 @@
+#define MCP2515_MODE_LOOPBACK           0x02
+#define MCP2515_MODE_SILENT             0x03
+#define MCP2515_MODE_CONFIGURATION      0x04
+
+/* MCP2515_FRAME_OFFSET */
+#define MCP2515_FRAME_OFFSET_SIDH       0
+#define MCP2515_FRAME_OFFSET_SIDL       1
+#define MCP2515_FRAME_OFFSET_EID8       2
+#define MCP2515_FRAME_OFFSET_EID0       3
+#define MCP2515_FRAME_OFFSET_DLC        4
+#define MCP2515_FRAME_OFFSET_D0         5
+
+/* MCP2515_CANINTF */
+#define MCP2515_CANINTF_RX0IF           BIT(0)
+#define MCP2515_CANINTF_RX1IF           BIT(1)
+#define MCP2515_CANINTF_TX0IF           BIT(2)
+#define MCP2515_CANINTF_TX1IF           BIT(3)
+#define MCP2515_CANINTF_TX2IF           BIT(4)
+#define MCP2515_CANINTF_ERRIF           BIT(5)
+#define MCP2515_CANINTF_WAKIF           BIT(6)
+#define MCP2515_CANINTF_MERRF           BIT(7)
+
+static int mcp2515_soft_reset(struct device *dev)
+{
+	u8_t opcode_buf[1] = { MCP2515_OPCODE_RESET };
+	const struct spi_buf tx_buf = {
+		.buf = opcode_buf,
+		.len = 1,
+	};
+	const struct spi_buf_set tx = {
+		.buffers = &tx_buf,
+		.count = 1
+	};
+
+	return spi_write(DEV_DATA(dev)->spi, &DEV_DATA(dev)->spi_cfg, &tx);
+}
+
+static int mcp2515_write_reg(struct device *dev, u8_t reg_addr, u8_t *buf_data,
+			     u8_t buf_len)
+{
+	u8_t opcode_buf[2];
+
+	opcode_buf[0] = MCP2515_OPCODE_WRITE;
+	opcode_buf[1] = reg_addr;
+
+	struct spi_buf tx_buf[2];
+
+	tx_buf[0].buf = opcode_buf;
+	tx_buf[0].len = 2;
+
+	tx_buf[1].buf = buf_data;
+	tx_buf[1].len = buf_len;
+
+	const struct spi_buf_set tx = {
+		.buffers = tx_buf,
+		.count = 2
+	};
+
+	return spi_write(DEV_DATA(dev)->spi, &DEV_DATA(dev)->spi_cfg, &tx);
+}
+
+static int mcp2515_bit_modify(struct device *dev, u8_t reg_addr, u8_t mask,
+			      u8_t data)
+{
+	u8_t cmd_buf[4];
+
+	cmd_buf[0] = MCP2515_OPCODE_BIT_MODIFY;
+	cmd_buf[1] = reg_addr;
+	cmd_buf[2] = mask;
+	cmd_buf[3] = data;
+
+	struct spi_buf tx_buf[1];
+
+	tx_buf[0].buf = cmd_buf;
+	tx_buf[0].len = 4;
+
+	const struct spi_buf_set tx = {
+		.buffers = tx_buf,
+		.count = 1
+	};
+
+	return spi_write(DEV_DATA(dev)->spi, &DEV_DATA(dev)->spi_cfg, &tx);
+}
+
+
+static int mcp2515_read_reg(struct device *dev, u8_t reg_addr, u8_t *buf_data,
+			    u8_t buf_len)
+{
+	u8_t opcode_buf[2];
+
+	opcode_buf[0] = MCP2515_OPCODE_READ;
+	opcode_buf[1] = reg_addr;
+
+	struct spi_buf tx_buf[2];
+
+	tx_buf[0].buf = opcode_buf;
+	tx_buf[0].len = 2;
+
+	tx_buf[1].buf = NULL;
+	tx_buf[1].len = buf_len;
+
+	const struct spi_buf_set tx = {
+		.buffers = tx_buf,
+		.count = 2
+	};
+
+	struct spi_buf rx_buf[2];
+
+	rx_buf[0].buf = NULL;
+	rx_buf[0].len = 2;
+
+	rx_buf[1].buf = buf_data;
+	rx_buf[1].len = buf_len;
+
+	const struct spi_buf_set rx = {
+		.buffers = rx_buf,
+		.count = 2
+	};
+
+	return spi_transceive(DEV_DATA(dev)->spi, &DEV_DATA(dev)->spi_cfg,
+			      &tx, &rx);
+}
+
+const int mcp2515_request_opmode(struct device *dev, u8_t mode)
+{
+	u8_t canstat;
+
+	mcp2515_bit_modify(dev, MCP2515_ADDR_CANCTRL, 0x07 << 5, mode << 5);
```

Comment:
0x07, 5: could you create macro to these with relevant names?

Reply:
I tried to do my best, there is currently 3 functions left where it is still using unnamed bit positions and masks
mcp2515_convert_can_msg_to_mcp2515_frame
mcp2515_convert_mcp2515_frame_to_can_msg
mcp2515_configure

They won't really become easier to read nor to follow, as you need to look at the datasheet if you want to know the details, and then it is easier to check for correctness when having the bare positions spilled out instead of double checking defines.
I can change the rest too if it is policy or just frowned upon.